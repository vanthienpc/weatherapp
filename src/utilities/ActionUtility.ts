import IAction from 'models/ActionModel';
import { call, put } from 'redux-saga/effects';
import ErrorResponseModel from 'models/ErrorResponseModel';

export function* createSagaEffect<P>(
  action: (...arg: any[]) => IAction<any>,
  effect: (...arg: any[]) => Promise<P | ErrorResponseModel>,
  ...args: any[]
): Generator {
  const response = yield call(effect, ...args);
  const isError: boolean = response instanceof ErrorResponseModel;

  return yield put(action(response, isError));
}

export const createAction = <T = undefined>(
  type: string,
  payload?: T,
  error = false,
  meta: any = null,
): IAction<T> => {
  return { type, payload, error, meta };
};
