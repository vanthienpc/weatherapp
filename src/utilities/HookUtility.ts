import {
  useSelector as originalUseSelector,
  useDispatch as originalUseDispatch,
  RootStateOrAny,
  TypedUseSelectorHook,
} from 'react-redux';

export const useSelector: TypedUseSelectorHook<RootStateOrAny> = (state: RootStateOrAny) =>
  originalUseSelector(state);
export const useDispatch = () => originalUseDispatch();
