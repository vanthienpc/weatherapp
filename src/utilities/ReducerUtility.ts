import { Reducer } from 'redux';
import IAction from 'models/ActionModel';

type ReducerMethod<T> = (state: T, action: IAction<any>) => T;
type ReducerMethods<T> = { [actionType: string]: ReducerMethod<T> };

export const handleReducer = <T = any>(initialState: T, methods: ReducerMethods<T>): Reducer<T> => {
  return (state: T = initialState, action: IAction<any>): T => {
    const method: ReducerMethod<T> | undefined = methods[action.type];

    if (!method || action.error) {
      return initialState;
    }

    return method(state, action);
  };
};
