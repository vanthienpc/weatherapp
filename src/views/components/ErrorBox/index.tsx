import React from 'react';
import { Result } from 'antd';
import IStore from 'models/IStore';
import { useSelector } from 'utilities/HookUtility';
import * as ErrorSelector from 'selectors/ErrorSelector';
import * as WeatherType from 'stores/weather/WeatherType';

const ErrorBox: React.FC = () => {
  const requestErrorText = useSelector((state: IStore) =>
    ErrorSelector.selectErrorText(state, [WeatherType.FETCH_WEATHER_FINISHED]),
  );

  if (requestErrorText) {
    return <Result status="warning" title={requestErrorText} />;
  }

  return null;
};

export default ErrorBox;
