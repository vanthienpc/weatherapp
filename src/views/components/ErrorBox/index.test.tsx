import React from 'react';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import createSagaMiddleware from 'redux-saga';
import configureStore from 'redux-mock-store';
import * as HookUtility from 'utilities/HookUtility';
import ErrorBox from '.';

const setup = (initialState?: string) => {
  let store: any;

  const sagaMiddleware = createSagaMiddleware();

  store = configureStore([sagaMiddleware])(initialState);

  jest.spyOn(HookUtility, 'useSelector').mockImplementation(() => store.getState());

  //@ts-ignore
  const wrapper = shallow(<ErrorBox store={store} />);

  return { wrapper, store };
};

it('render should match the snapshot', () => {
  const initialState = 'city not found';
  const { wrapper } = setup(initialState);
  expect(toJson(wrapper)).toMatchSnapshot();
});
