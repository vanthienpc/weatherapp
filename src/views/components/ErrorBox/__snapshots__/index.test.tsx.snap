// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render should match the snapshot 1`] = `
<Result
  status="warning"
  title="city not found"
/>
`;
